File: dart/ftr_verify/lib/src/canonical_cbor.dart
/// Canonical CBOR — deterministic encoding, RFC 8949 §4.2.1 core rules.
///
/// This is a **second, independent implementation** of the profile in
/// `core/ftr/canonical_cbor.py`. That is the entire point of it. The digest of a
/// Field Test Record is what a §63 certificate must state, and a digest is only
/// worth stating if two people who wrote two encoders separately arrive at the
/// same bytes. A single implementation agreeing with itself proves nothing.
///
/// The profile, restated here so this file can be read without the Python:
///
///   * definite lengths only — no streaming forms
///   * shortest-form integer and length encoding
///   * map keys sorted bytewise-lexicographically by their *encoded* bytes
///   * no floats, no tags, no duplicate keys
///
/// Floats are refused rather than canonicalised: every measured quantity in a
/// record is a scaled integer with its scale named in the field.
library;

import 'dart:convert';
import 'dart:typed_data';

class CborError implements Exception {
  final String message;
  CborError(this.message);
  @override
  String toString() => 'CborError: $message';
}

// --------------------------------------------------------------------------- //
// encoding
// --------------------------------------------------------------------------- //

void _writeHead(BytesBuilder out, int major, int n) {
  if (n < 0) throw CborError('negative argument $n');
  final m = major << 5;
  if (n < 24) {
    out.addByte(m | n);
  } else if (n < 0x100) {
    out..addByte(m | 24)..addByte(n);
  } else if (n < 0x10000) {
    out.addByte(m | 25);
    out.add([(n >> 8) & 0xff, n & 0xff]);
  } else if (n < 0x100000000) {
    out.addByte(m | 26);
    out.add([(n >> 24) & 0xff, (n >> 16) & 0xff, (n >> 8) & 0xff, n & 0xff]);
  } else {
    // Dart ints are 64-bit signed. Values at or above 2^63 cannot be represented,
    // and nothing in a record needs them; refusing beats silently wrapping.
    if (n < 0) throw CborError('integer out of range');
    out.addByte(m | 27);
    for (var s = 56; s >= 0; s -= 8) {
      out.addByte((n >> s) & 0xff);
    }
  }
}

void _encodeInto(BytesBuilder out, Object? v) {
  if (v == null) {
    out.addByte(0xf6);
  } else if (v is bool) {
    out.addByte(v ? 0xf5 : 0xf4);
  } else if (v is int) {
    if (v >= 0) {
      _writeHead(out, 0, v);
    } else {
      _writeHead(out, 1, -v - 1);
    }
  } else if (v is double) {
    throw CborError('floats are not encodable in an FTR — store a scaled integer '
        'and name the scale in the field (e.g. delta_e_x1000)');
  } else if (v is Uint8List) {
    // Only Uint8List is a byte string. A plain List<int> encodes as an array of
    // integers, which mirrors Python's bytes-versus-list distinction exactly — the
    // two implementations must not disagree about what a value *is*.
    _writeHead(out, 2, v.length);
    out.add(v);
  } else if (v is String) {
    final b = utf8.encode(v);
    _writeHead(out, 3, b.length);
    out.add(b);
  } else if (v is List) {
    _writeHead(out, 4, v.length);
    for (final x in v) {
      _encodeInto(out, x);
    }
  } else if (v is Map) {
    final items = <MapEntry<Uint8List, Uint8List>>[];
    v.forEach((k, val) {
      if (k is! String && k is! int) {
        throw CborError('map keys must be String or int, got ${k.runtimeType}');
      }
      items.add(MapEntry(encode(k), encode(val)));
    });
    items.sort((a, b) => _compareBytes(a.key, b.key));
    for (var i = 1; i < items.length; i++) {
      if (_compareBytes(items[i - 1].key, items[i].key) == 0) {
        throw CborError('duplicate map key after encoding');
      }
    }
    _writeHead(out, 5, items.length);
    for (final e in items) {
      out..add(e.key)..add(e.value);
    }
  } else {
    throw CborError('type not encodable in an FTR: ${v.runtimeType}');
  }
}

int _compareBytes(Uint8List a, Uint8List b) {
  final n = a.length < b.length ? a.length : b.length;
  for (var i = 0; i < n; i++) {
    if (a[i] != b[i]) return a[i] - b[i];
  }
  return a.length - b.length;
}

/// Encode [value] canonically.
Uint8List encode(Object? value) {
  final out = BytesBuilder(copy: false);
  _encodeInto(out, value);
  return out.toBytes();
}

// --------------------------------------------------------------------------- //
// decoding
// --------------------------------------------------------------------------- //

class _Reader {
  final Uint8List buf;
  int i = 0;
  _Reader(this.buf);

  (int, int) readHead() {
    if (i >= buf.length) throw CborError('truncated: expected a type byte');
    final ib = buf[i++];
    final major = ib >> 5;
    final minor = ib & 0x1f;
    if (minor < 24) return (major, minor);
    if (minor == 31) throw CborError('indefinite-length item is not canonical');
    if (minor > 27) throw CborError('reserved additional-information value $minor');
    final width = 1 << (minor - 24);
    if (i + width > buf.length) {
      throw CborError('truncated: argument runs past end of buffer');
    }
    var n = 0;
    for (var k = 0; k < width; k++) {
      n = (n << 8) | buf[i + k];
    }
    i += width;
    if (n < 0) throw CborError('integer too large for this implementation');
    // Shortest-form check: a longer encoding of a small value is not canonical.
    const limits = {1: 24, 2: 0x100, 4: 0x10000, 8: 0x100000000};
    if (n < limits[width]!) {
      throw CborError('non-canonical: $n encoded in $width byte(s), shorter form exists');
    }
    return (major, n);
  }

  Object? read() {
    final (major, n) = readHead();
    switch (major) {
      case 0:
        return n;
      case 1:
        return -1 - n;
      case 2:
        if (i + n > buf.length) throw CborError('truncated byte string');
        final b = Uint8List.sublistView(buf, i, i + n);
        i += n;
        return Uint8List.fromList(b);
      case 3:
        if (i + n > buf.length) throw CborError('truncated text string');
        final b = Uint8List.sublistView(buf, i, i + n);
        i += n;
        try {
          return utf8.decode(b);
        } on FormatException {
          throw CborError('invalid UTF-8 in text string');
        }
      case 4:
        return List<Object?>.generate(n, (_) => read(), growable: false);
      case 5:
        final out = <Object?, Object?>{};
        Uint8List? prev;
        for (var k = 0; k < n; k++) {
          final start = i;
          final key = read();
          final enc = Uint8List.fromList(Uint8List.sublistView(buf, start, i));
          if (prev != null && _compareBytes(enc, prev) <= 0) {
            throw CborError('non-canonical: map keys out of order or duplicated');
          }
          prev = enc;
          out[key] = read();
        }
        return out;
      case 7:
        if (n == 20) return false;
        if (n == 21) return true;
        if (n == 22) return null;
        throw CborError('simple/float value $n is not permitted in an FTR');
      default:
        throw CborError('unsupported major type $major');
    }
  }
}

/// Decode [buf], rejecting anything a canonical encoder could not have produced.
Object? decode(Uint8List buf) {
  final r = _Reader(buf);
  final v = r.read();
  if (r.i != buf.length) {
    throw CborError('${buf.length - r.i} trailing byte(s) after the top-level item');
  }
  return v;
}

/// True if [buf] decodes and re-encodes to itself, byte for byte.
bool isCanonical(Uint8List buf) {
  try {
    return _compareBytes(encode(decode(buf)), buf) == 0;
  } on CborError {
    return false;
  }
}

File: core/ftr/canonical_cbor.py
"""Canonical CBOR — deterministic encoding, RFC 8949 §4.2.1 core rules.

The digest is the legal artefact, so serialisation must be byte-identical for
identical content on every implementation, forever. That rules out most of CBOR:

  * definite lengths only — no streaming forms
  * shortest-form integer and length encoding
  * map keys sorted bytewise-lexicographically by their *encoded* bytes
  * no floats, no tags, no duplicate keys

Floats are rejected outright rather than canonicalised. IEEE-754 rounding is not
worth defending under cross-examination; every measured quantity in an FTR is
stored as a scaled integer with its scale named in the field (``delta_e_x1000``).

This module has no third-party dependencies on purpose: the verifier must be
runnable by someone who does not trust us and will not install our packages.
"""

from __future__ import annotations

from typing import Any

__all__ = ["dumps", "loads", "is_canonical", "CborError"]


class CborError(ValueError):
    """Raised for input this profile refuses to encode or decode."""


# --------------------------------------------------------------------------- #
# encoding
# --------------------------------------------------------------------------- #

def _head(major: int, n: int) -> bytes:
    """Type byte plus shortest-form argument for ``n``."""
    m = major << 5
    if n < 24:
        return bytes([m | n])
    if n < 0x100:
        return bytes([m | 24, n])
    if n < 0x10000:
        return bytes([m | 25]) + n.to_bytes(2, "big")
    if n < 0x100000000:
        return bytes([m | 26]) + n.to_bytes(4, "big")
    if n < 0x10000000000000000:
        return bytes([m | 27]) + n.to_bytes(8, "big")
    raise CborError(f"integer too large for CBOR: {n}")


def _encode(v: Any) -> bytes:
    # bool before int: bool is an int subclass in Python and would encode as 0/1.
    if v is True:
        return b"\xf5"
    if v is False:
        return b"\xf4"
    if v is None:
        return b"\xf6"
    if isinstance(v, int):
        return _head(0, v) if v >= 0 else _head(1, -v - 1)
    if isinstance(v, float):
        raise CborError(
            "floats are not encodable in an FTR — store a scaled integer and "
            "name the scale in the field (e.g. delta_e_x1000)"
        )
    if isinstance(v, bytes):
        return _head(2, len(v)) + v
    if isinstance(v, str):
        b = v.encode("utf-8")
        return _head(3, len(b)) + b
    if isinstance(v, (list, tuple)):
        return _head(4, len(v)) + b"".join(_encode(x) for x in v)
    if isinstance(v, dict):
        items = []
        for k, val in v.items():
            if not isinstance(k, (str, int)) or isinstance(k, bool):
                raise CborError(f"map keys must be str or int, got {type(k).__name__}")
            items.append((_encode(k), _encode(val)))
        keys = [k for k, _ in items]
        if len(set(keys)) != len(keys):
            raise CborError("duplicate map key after encoding")
        items.sort(key=lambda kv: kv[0])  # bytewise lexicographic on encoded keys
        return _head(5, len(items)) + b"".join(k + val for k, val in items)
    raise CborError(f"type not encodable in an FTR: {type(v).__name__}")


def dumps(value: Any) -> bytes:
    """Encode ``value`` canonically."""
    return _encode(value)


# --------------------------------------------------------------------------- #
# decoding
# --------------------------------------------------------------------------- #

def _read_head(buf: bytes, i: int) -> tuple[int, int, int]:
    if i >= len(buf):
        raise CborError("truncated: expected a type byte")
    ib = buf[i]
    major, minor = ib >> 5, ib & 0x1F
    i += 1
    if minor < 24:
        return major, minor, i
    if minor == 31:
        raise CborError("indefinite-length item is not canonical")
    if minor > 27:
        raise CborError(f"reserved additional-information value {minor}")
    width = 1 << (minor - 24)
    if i + width > len(buf):
        raise CborError("truncated: argument runs past end of buffer")
    n = int.from_bytes(buf[i:i + width], "big")
    # shortest-form check — a longer encoding of a small value is not canonical
    limits = {1: 24, 2: 0x100, 4: 0x10000, 8: 0x100000000}
    if n < limits[width]:
        raise CborError(f"non-canonical: {n} encoded in {width} byte(s), shorter form exists")
    return major, n, i + width


def _decode(buf: bytes, i: int) -> tuple[Any, int]:
    major, n, i = _read_head(buf, i)
    if major == 0:
        return n, i
    if major == 1:
        return -1 - n, i
    if major == 2:
        if i + n > len(buf):
            raise CborError("truncated byte string")
        return buf[i:i + n], i + n
    if major == 3:
        if i + n > len(buf):
            raise CborError("truncated text string")
        try:
            return buf[i:i + n].decode("utf-8"), i + n
        except UnicodeDecodeError as e:
            raise CborError("invalid UTF-8 in text string") from e
    if major == 4:
        out = []
        for _ in range(n):
            v, i = _decode(buf, i)
            out.append(v)
        return out, i
    if major == 5:
        out: dict[Any, Any] = {}
        prev: bytes | None = None
        for _ in range(n):
            start = i
            k, i = _decode(buf, i)
            enc = buf[start:i]
            if prev is not None and enc <= prev:
                raise CborError("non-canonical: map keys out of order or duplicated")
            prev = enc
            v, i = _decode(buf, i)
            out[k] = v
        return out, i
    if major == 7:
        if n == 20:
            return False, i
        if n == 21:
            return True, i
        if n == 22:
            return None, i
        raise CborError(f"simple/float value {n} is not permitted in an FTR")
    raise CborError(f"unsupported major type {major}")


def loads(buf: bytes) -> Any:
    """Decode ``buf``, rejecting anything a canonical encoder could not have produced."""
    if not isinstance(buf, (bytes, bytearray)):
        raise CborError("loads expects bytes")
    v, i = _decode(bytes(buf), 0)
    if i != len(buf):
        raise CborError(f"{len(buf) - i} trailing byte(s) after the top-level item")
    return v


def is_canonical(buf: bytes) -> bool:
    """True if ``buf`` decodes and re-encodes to itself, byte for byte."""
    try:
        return dumps(loads(buf)) == bytes(buf)
    except CborError:
        return False
